totalRP3_Extended/aura/aura.lua

```diff
@@ -0,0 +1,1009 @@
+				end
+			end
+			
+			-- part 4: do it again in case there were workflows running
+			if numScripts <= 0 then
+				updateCycle = MAX_AURA_UPDATE_CYCLES + 1; -- get out of the loop
+			end
+			
+		end
+		
+		if aurasRemoved then
+			needsHardRefresh = true;
+			self:RemoveInvalidAuras();
+		end
+		
+		local nextTimestamp = math.huge;
+		
+		for i = 1,#self.activeAuras do
+			local aura = self.activeAuras[i];
+			aura.persistent.dormantSince = now;
+			if aura.class.BA.OV then
+				aura.overlay = TRP3_API.script.parseArgs(aura.class.BA.OV, { object = aura.persistent });
+			else
+				aura.overlay = nil;
+			end
+			
+			if aura.persistent.expiry < math.huge then
+				aura.duration = self.timeFormatterCompact:Format(aura.persistent.expiry - now);
+			else
+				aura.duration = nil;
+			end
+			
+			if aura.class.BA.IV and aura.class.BA.IV < math.huge then
+				local nextTick = aura.persistent.lastTick + math.abs(aura.class.BA.IV);
+				nextTimestamp = math.min(nextTimestamp, nextTick);
+			end
+			
+			if aura.hasDynamicOverlay then
+				nextTimestamp = math.min(nextTimestamp, now + DYN_AURA_UPDATE_INTERVAL);
+			end
+			
+			if aura.persistent.expiry < math.huge then
+				nextTimestamp = math.min(nextTimestamp, aura.persistent.expiry);
+				local timeTillExpiry = aura.persistent.expiry - now;
+				if timeTillExpiry < 60 then
+					nextTimestamp = math.min(nextTimestamp, aura.persistent.expiry - math.floor(timeTillExpiry));
+				elseif timeTillExpiry < 3600 then
+					nextTimestamp = math.min(nextTimestamp, aura.persistent.expiry - math.floor(timeTillExpiry/60) * 60);
+				else
+					nextTimestamp = math.min(nextTimestamp, aura.persistent.expiry - math.floor(timeTillExpiry/3600) * 3600);
+				end
+			end
+			
+		end
+		
+		if needsHardRefresh then
+			self:HardRefresh();
+		else
+			self:SoftRefresh();
+		end
+		
+		if nextTimestamp < math.huge then
+			self.timer = C_Timer.NewTimer(math.max(nextTimestamp - now, MIN_AURA_UPDATE_INTERVAL), function() 
+				auraCore:Update();
+			end)
+		end
+		
+		self.isUpdating = false;
+	end,
+	
+	RemoveInvalidAuras = function(self)
+		-- 1st pass: remove from active auras list
+		local n = #self.activeAuras;
+		local j = 1;
+		for i = 1,n do
+			if self.activeAuras[i].persistent.invalid then
+				self.activeAuras[i] = nil;
+			else
+				if i ~= j then
+					self.activeAuras[j] = self.activeAuras[i];
+					self.activeAuras[i] = nil;
+				end
+				j = j + 1;
+			end
+		end
+		-- 2nd pass: remove from all auras list
+		local n = #self.currentProfile.auras;
+		local j = 1;
+		for i = 1,n do
+			if self.currentProfile.auras[i].invalid then
+				self.currentProfile.auras[i] = nil;
+			else
+				if i ~= j then
+					self.currentProfile.auras[j] = self.currentProfile.auras[i];
+					self.currentProfile.auras[i] = nil;
+				end
+				j = j + 1;
+			end
+		end
+	end,
+	
+	-- updates the aura display entirely, re-assigning frames etc.
+	-- do this, when auras are added or removed
+	HardRefresh = function(self)
```

Comment:
It's 2023 and people still aren't using the grid layout utilities :sadge:

A lot of this frame initialization and layout logic can be automated.

```lua
-- Tweak these as needed.
-- You'd need to repeat this separately for buffs and debuffs; this example
-- would just be for buffs for now.
local initialAnchor = AnchorUtil.CreateAnchor("TOPRIGHT", TRP3_AuraFrame, "TOPRIGHT", 0, 0);
local direction = GridLayoutMixin.Direction.TopRightToBottomLeft;
local stride = AURA_ROW_LENGTH;
local paddingX = 0;
local paddingY = 0;
local layout = AnchorUtil.CreateGridLayout(direction, stride, paddingX, paddingY);

local buffs = { TODO };

local function BuffButtonFactory(index)
    local frame = self.auraFramePool:Acquire(index);
    local aura = buffs[index];
    frame:Init(aura);  -- Move logic like SetTexture/SetVertexCalls into an Init method on the buttons.
    aura.frame = frame;
    return frame;
end;

local buffCount = #buffs;
local buffButtons = AnchorUtil.GridLayoutFactoryByCount(BuffButtonFactory, buffCount, initialAnchor, layout);
```